components/movementsensor/mpu6050/mpu6050.go

```diff
@@ -0,0 +1,328 @@
+	// To do this, we set register 107 to 0.
+	err = sensor.writeByte(ctx, 107, 0)
+	if err != nil {
+		return nil, errors.Errorf("Unable to wake up MPU6050: '%s'", err.Error())
+	}
+
+	// Now, turn on the background goroutine that constantly reads from the chip and stores data in
+	// the object we created.
+	sensor.activeBackgroundWorkers.Add(1)
+	utils.PanicCapturingGo(sensor.pollData)
+
+	return sensor, nil
+}
+
+func (mpu *mpu6050) readByte(ctx context.Context, register byte) (byte, error) {
+	result, err := mpu.readBlock(ctx, register, 1)
+	if err != nil {
+		return 0, err
+	}
+	return result[0], err
+}
+
+func (mpu *mpu6050) readBlock(ctx context.Context, register byte, length uint8) ([]byte, error) {
+	handle, err := mpu.bus.OpenHandle(mpu.i2cAddress)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err := handle.Close()
+		if err != nil {
+			mpu.logger.Error(err)
+		}
+	}()
+
+	results, err := handle.ReadBlockData(ctx, register, length)
+	return results, err
+}
+
+func (mpu *mpu6050) writeByte(ctx context.Context, register, value byte) error {
+	handle, err := mpu.bus.OpenHandle(mpu.i2cAddress)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err := handle.Close()
+		if err != nil {
+			mpu.logger.Error(err)
+		}
+	}()
+
+	return handle.WriteByteData(ctx, register, value)
+}
+
+// A helper function: takes 2 bytes and reinterprets them as a big-endian signed integer.
+func toSignedValue(data []byte) int {
+	return int(int16(binary.BigEndian.Uint16(data)))
+}
+
+// Given a value, scales it so that the range of int16s becomes the range of +/- maxValue.
+func setScale(value int, maxValue float64) float64 {
+	return float64(value) * maxValue / (1 << 15)
+}
+
+// A helper function to abstract out shared code: takes 6 bytes and gives back AngularVelocity.
+func toAngularVelocity(data []byte) spatialmath.AngularVelocity {
+	gx := toSignedValue(data[0:2])
+	gy := toSignedValue(data[2:4])
+	gz := toSignedValue(data[4:6])
+
+	maxRotation := 250.0 // Degrees per second
+	return spatialmath.AngularVelocity{
+		X: setScale(gx, maxRotation),
+		Y: setScale(gy, maxRotation),
+		Z: setScale(gz, maxRotation),
+	}
+}
+
+// A helper function that takes 6 bytes and gives back linear acceleration.
+func toLinearAcceleration(data []byte) r3.Vector {
+	x := toSignedValue(data[0:2])
+	y := toSignedValue(data[2:4])
+	z := toSignedValue(data[4:6])
+
+	// The scale is +/- 2G's, but our units should be mm/sec/sec.
+	maxAcceleration := 2.0 * 9.81 /* m/sec/sec */ * 1000.0 /* mm/m */
+	return r3.Vector{
+		X: setScale(x, maxAcceleration),
+		Y: setScale(y, maxAcceleration),
+		Z: setScale(z, maxAcceleration),
+	}
+}
+
+func (mpu *mpu6050) pollData() {
+	defer mpu.activeBackgroundWorkers.Done()
+
+	for mpu.backgroundContext.Err() == nil {
+		// Reading data and re-checking for context errors once per millisecond is probably fast
+		// enough.
+		time.Sleep(time.Millisecond)
```

Comment:
You can also set a timer and send the ticker data to a channel. Not necessary here but it's a cool go pattern. See the example in `imu-vectornav''s `init` function.

Reply:
That version is much nicer: mine will have up to a 1ms delay when shutting down, whereas yours will not. I'll make that change, too...